File: java/array_search/FibonacciSearch.java
public class FibonacciSearch {

    public static int fibonacciSearch(int[] arr, int target) {
        int n = arr.length;

        if (n == 0) return -1;

        int fibMm2 = 0;
        int fibMm1 = 1;
        int fibM = fibMm2 + fibMm1;

        while (fibM < n) {
            fibMm2 = fibMm1;
            fibMm1 = fibM;
            fibM = fibMm2 + fibMm1;
        }

        int offset = -1;

        while (fibM > 1) {
            int index = Math.min(offset + fibMm2, n - 1);

            if (arr[index] < target) {
                fibM = fibMm1;
                fibMm1 = fibMm2;
                fibMm2 = fibM - fibMm1;
                offset = index;
            } else if (arr[index] > target) {
                fibM = fibMm2;
                fibMm1 -= fibMm2;
                fibMm2 = fibM - fibMm1;
            } else {
                return index;
            }
        }

        if (fibMm1 == 1 && offset + 1 < n && arr[offset + 1] == target) {
            return offset + 1;
        }

        return -1;
    }
}

File: c/array_search/fibonacci_search.c
int fibonacci_search(int *arr, int n, int target)
{
    if (n <= 0) return -1;

    int fib_mm2 = 0;
    int fib_mm1 = 1;
    int fib_m = fib_mm2 + fib_mm1;

    while (fib_m < n)
    {
        fib_mm2 = fib_mm1;
        fib_mm1 = fib_m;
        fib_m = fib_mm2 + fib_mm1;
    }

    int offset = -1;

    while (fib_m > 1)
    {
        int index = offset + fib_mm2;
        if (index >= n) index = n - 1;

        if (arr[index] < target)
        {
            fib_m = fib_mm1;
            fib_mm1 = fib_mm2;
            fib_mm2 = fib_m - fib_mm1;
            offset = index;
        }
        else if (arr[index] > target)
        {
            fib_m = fib_mm2;
            fib_mm1 = fib_mm1 - fib_mm2;
            fib_mm2 = fib_m - fib_mm1;
        }
        else
        {
            return index;
        }
    }

    if (fib_mm1 && offset + 1 < n && arr[offset + 1] == target)
    {
        return offset + 1;
    }

    return -1;
}
